content/en/BestPractices/akhan_input_validation.md

```diff
@@ -0,0 +1,155 @@
+---
+title: Validate User Input
+linkTitle: Validate User Input
+weight: 610
+description: >
+  Here's an attack scenario that could cost you money
+---
+
+{{% pageinfo %}}
+Protect your systems from bad user input. In this article, we share best
+practices to validate user input, securely.
+{{% /pageinfo %}}
+
+To create secure software, you need code that checks user input. Without such
+checks, you could see security vulnerabilities such as:
+
+- [Business logic flaws](https://owasp.org/www-community/vulnerabilities/Business_logic_vulnerability)
+- [Denial of Service attacks](https://en.wikipedia.org/wiki/Denial-of-service_attack)
+- [Injection
+  flaws](https://owasp.org/www-community/Injection_Flaws#:~:text=An%20injection%20flaw%20is%20a,connected%20to%20the%20vunlerable%20application.)
+
+In this article, we explore an attack scenario. We also show how you can minimize your risks. 
+Read further to learn best practices to safely validate user input.
+
+## Description
+
+Developers who write secure code use [Input
+Validation](https://cwe.mitre.org/data/definitions/20.html). Black hat hackers
+frequently attack websites with questionable input.
+
+## Attack Scenario
+
+An attacker might think: "I can earn free money. All I need to do is find a
+website that allows me to buy a negative quantity of products."
+
+Review this code sample from the [OWASP Juice Shop](https://github.com/juice-shop/juice-shop/).
+It's a perfect example of how an innocent mistake can cause huge business losses:	
+
+```ts
+ async function quantityCheck (req: Request, res: Response, next: NextFunction, id: number, quantity: number) {
+  const product = await QuantityModel.findOne({ where: { ProductId: id } })
+  if (!product) {
+    throw new Error('No such product found!')
+  }
+
+  if (!product.limitPerUser || (product.limitPerUser && product.limitPerUser >= quantity) || security.isDeluxe(req)) {
+    if (product.quantity >= quantity) { // enough in stock?
+      next()
+    } else {
+      res.status(400).json({ error: res.__('We are out of stock! Sorry for the inconvenience.') })
+    }
+  } else {
+    res.status(400).json({ error: res.__('You can order only up to {{quantity}} items of this product.', { quantity: product.limitPerUser.toString() }) })
+  }
+}
+```
+
+### Analysis
+
+The OWASP Juice Shop uses an [API
+Post](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/server.ts#L370)
+to call a
+[`quantityCheck()`](https://github.com/juice-shop/juice-shop/blob/bd7e8f3c7af1a8f38ba4fa866000136cb472449a/routes/basketItems.ts#L90)
+function.
+
+It checks the quantity in the request payload with validation logic, using
+`product.limitPerUser`. However, it does not check if the quantity exceeds the:
```

Comment:
The `quantityCheck()` function checks if the quantity is valid and doesn't exceed the limit allowed per user or, the limit allowed per order. It bypasses the check if the user is a `deluxe` user.

Reply:
See [next comment](https://github.com/cobalthq/cobalt-product-public-docs/pull/154/files#r945688933) for response